File: logs-deleter/lib/logsDeleter.go
package lib

import (
	"log"
	"os"
	"path/filepath"
	"time"
)

func DeleteFrom(subPath string, days int, deleteEmpty bool, verbose bool) error {
	cutoff := time.Now().AddDate(0, 0, -1*days)

	if verbose {
		log.Printf("Searching %v for files older than %v\n", subPath, cutoff)
	}

	files, err := filepath.Glob(filepath.Join(subPath, "*"))
	if err != nil {
		return err
	}

	filesToDelete := make([]os.FileInfo, 0)
	for _, file := range files {
		fileStat, err := os.Stat(file)
		if err != nil {
			return err
		}
		if fileStat.ModTime().Before(cutoff) {
			filesToDelete = append(filesToDelete, fileStat)
		} else if deleteEmpty && fileStat.Size() == 0 {
			filesToDelete = append(filesToDelete, fileStat)
		}
	}

	countFilesToDelete := len(filesToDelete)

	if verbose || countFilesToDelete > 0 {
		log.Printf("Found %d files to delete in %v\n", len(filesToDelete), subPath)
	}

	for _, fileToDelete := range filesToDelete {
		var pathToDelete = filepath.Join(subPath, fileToDelete.Name())
		log.Printf("Deleting %v\n", pathToDelete)

		err = os.RemoveAll(pathToDelete)
		if err != nil {
			return err
		}
	}

	return nil
}

File: LogsDeleter/LogsDeleter/LogsDeleter.cs
using Microsoft.Extensions.Logging;

namespace LogsDeleter;

using System;
using System.Collections.Generic;
using System.IO;

public class LogsDeleter
{
    private readonly ILogger<LogsDeleter> _logger;

    public LogsDeleter(ILogger<LogsDeleter> logger)
    {
        ArgumentNullException.ThrowIfNull(logger);

        _logger = logger;
    }

    public void DeleteFrom(string subPath, int days, bool deleteEmpty, bool verbose)
    {
        if (days <= 0)
        {
            throw new ArgumentException($"days must be positive, got: {days}", nameof(days));
        }

        // Go's AddDate(0, 0, -days) equivalent
        DateTime cutoff = DateTime.Now.AddDays(-days);

        if (verbose)
        {
            _logger.LogInformation($"Searching {subPath} for files older than {cutoff}");
        }

        if (!Directory.Exists(subPath))
        {
            throw new DirectoryNotFoundException($"The path {subPath} does not exist.");
        }

        // Using EnumerateFiles to get full paths, similar to filepath.Glob
        var allFiles = Directory.EnumerateFiles(subPath, "*");
        var filesToDelete = new List<FileInfo>();

        foreach (var filePath in allFiles)
        {
            var fileInfo = new FileInfo(filePath);

            // Logic: Older than cutoff OR (deleteEmpty is true AND file is 0 bytes)
            if (fileInfo.LastWriteTime < cutoff)
            {
                filesToDelete.Add(fileInfo);
            }
            else if (deleteEmpty && fileInfo.Length == 0)
            {
                filesToDelete.Add(fileInfo);
            }
        }

        if (verbose || filesToDelete.Count > 0)
        {
            _logger.LogInformation($"Found {filesToDelete.Count} files to delete in {subPath}");
        }

        foreach (var file in filesToDelete)
        {
            _logger.LogInformation($"Deleting {file.FullName}");

            try
            {
                // File.Delete is the equivalent of os.Remove
                // If you need to delete directories too, use Directory.Delete(path, true)
                file.Delete();
            }
            catch (Exception ex)
            {
                // Go returns the error; C# typically throws.
                // Wrapping in a custom exception to match the "return err" flow.
                throw new IOException($"Failed to delete {file.FullName}", ex);
            }
        }
    }
}
